src/vcf/fixer.cpp

```diff
@@ -210,10 +243,139 @@ namespace ebi
+                fixed_column = empty_value;
+            } else {
+                fixed_column.pop_back();          // remove trailing separator (its length is always 1)
+            }
+
+            output << fixed_column;
+            return num_removed_duplicates;
+        }
+
+        size_t Fixer::remove_duplicate_format_sample_pairs(const std::string &string_line) {
+            const size_t format_column_index = 8;
+            size_t num_removed_duplicates = 0;
+
+            using iter = std::vector<std::string>::iterator;
+            fix_columns(format_column_index, -1, string_line, "\t", [&](iter first, iter last) {
+                std::map<std::string, std::vector<size_t>> format_fields;
+                std::vector<std::string> ordered_fields;
+                std::vector<std::string> format_keys;
+                util::string_split(*first, ":", format_keys);
+
+                for (size_t i = 0; i < format_keys.size(); i++) {
+                    format_fields[format_keys[i]].push_back(i);
+                    if (format_fields[format_keys[i]].size() == 1) {
+                        ordered_fields.push_back(format_keys[i]);
+                    }
+                }
+
+                std::set<std::string> fields_to_remove;
+
+                for (auto & format_field : format_fields) {
+                    if (format_field.second.size() > 1) {
+                        for (auto it = first + 1; it != last; it++) {
+                            std::vector<std::string> sample_fields;
+                            util::string_split(*it, ":", sample_fields);
+                            for (auto & format_field_index : format_field.second) {
+                                if (sample_fields.size() > format_field_index && sample_fields[format_field_index] != sample_fields[format_field.second[0]]) {
+                                    fields_to_remove.insert(format_field.first);
+                                    it = last - 1;
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+
+                std::string fixed_format;
+                std::vector<std::string> fixed_samples(last - (first + 1));
+
+                for (auto & ordered_field : ordered_fields) {
+                    if (fields_to_remove.find(ordered_field) == fields_to_remove.end()) {
+                        // keep first occurrence, discard the rest if present
+                        fixed_format += ordered_field + ":";
+                        for (auto it = first + 1; it != last; it++) {
+                            std::vector<std::string> sample_fields;
+                            util::string_split(*it, ":", sample_fields);
+                            if (sample_fields.size() > format_fields[ordered_field][0]) {
+                                fixed_samples[it - (first + 1)] += sample_fields[format_fields[ordered_field][0]] + ":";
```

Comment:
this concatenation might be easier using util::print_container, because it accepts the separator, and it handles it so it's not necessary to do the pop_back later